### ISSUE TITLE
Missing 'week' section in remove_old_partitions

### ISSUE THREAD

**Issue by beinvisible** [OpensourceICTSolutions/zabbix-mysql-partitioning-perl#4]
seems like using 'week' will always remove *all* partitions, because in sub remove_old_partitions the date is not decremented at all.

**Comment by larcorba** [comment]
Update: 
**Workaround. For now the script is usable by keeping the _daily_ partition for history and _monthly_ for trends.**


Hi Beinvisble,

What database type and version are you running? Have you actually noticed this behaviour in your environment? 

Seems like it's working fine in my environments and the 'period' is used correctly.

Let me know.

Regards,

Nathan

**Comment by beinvisible** [comment]
For 'week' the executed code in remove_old_partitions is:

> my $curr_date = DateTime->now;
> ...
> elsif ( $period eq 'week' ) {
>     # empty
> }
> ...
> foreach my $partition (sort keys %{$table_part}) {
>     if ($table_part->{$partition}->{'partition_description'} <= $curr_date->epoch) {
>         my $query = "ALTER TABLE $db_schema.$table_name DROP PARTITION $partition";
>         $dbh->do($query);
>     }
> }

i.e. it obviously drops all partitions up to 'now', because 'partition_description' is the timestamp as defined in "ADD PARTITION ... VALUES LESS THAN (timestamp)".

**Comment by larcorba** [comment]
Just tested with a 30 day retention period:

> Sep 10 10:40:58 lar-permanent-test-5x0 mysql_zbx_part[626323]: Removing old p2021_08_09 partition from history table
> Sep 10 10:40:58 lar-permanent-test-5x0 mysql_zbx_part[626323]: ALTER TABLE zabbix.history DROP PARTITION p2021_08_09
> Sep 10 10:40:58 lar-permanent-test-5x0 mysql_zbx_part[626323]: Removing old p2021_08_10 partition from history table
> Sep 10 10:40:58 lar-permanent-test-5x0 mysql_zbx_part[626323]: ALTER TABLE zabbix.history DROP PARTITION p2021_08_10

Oldest partition deleted is 2021_08_10, history is kept in database as expected.

Checked table:
> SELECT * FROM history PARTITION (p2021_09_08);

Got results:
> |  23277 | 1631138397 |     1.5330211527638877 | 861162674 |
> |  10077 | 1631138397 |    0.13330585836050063 | 861234040 |
> |  10078 | 1631138398 |    0.16663276058247214 | 861457185 |
> +--------+------------+------------------------+-----------+


Changed retention to two days:

> Removing old p2021_09_04 partition from history table
> Sep 10 10:57:54 lar-permanent-test-5x0 mysql_zbx_part[628275]: ALTER TABLE zabbix.history DROP PARTITION p2021_09_04
> Sep 10 10:57:54 lar-permanent-test-5x0 mysql_zbx_part[628275]: Removing old p2021_09_05 partition from history table
> Sep 10 10:57:54 lar-permanent-test-5x0 mysql_zbx_part[628275]: ALTER TABLE zabbix.history DROP PARTITION p2021_09_05
> Sep 10 10:57:54 lar-permanent-test-5x0 mysql_zbx_part[628275]: Removing old p2021_09_06 partition from history table
> Sep 10 10:57:54 lar-permanent-test-5x0 mysql_zbx_part[628275]: ALTER TABLE zabbix.history DROP PARTITION p2021_09_06
> Sep 10 10:57:54 lar-permanent-test-5x0 mysql_zbx_part[628275]: Removing old p2021_09_07 partition from history table
> Sep 10 10:57:54 lar-permanent-test-5x0 mysql_zbx_part[628275]: ALTER TABLE zabbix.history DROP PARTITION p2021_09_07


Again:
> SELECT * FROM history PARTITION (p2021_09_08);


Got results:

> |  23277 | 1631138397 |     1.5330211527638877 | 861162674 |
> |  10077 | 1631138397 |    0.13330585836050063 | 861234040 |
> |  10078 | 1631138398 |    0.16663276058247214 | 861457185 |
> +--------+------------+------------------------+-----------+

Script. seems to be working just fine :)

**Comment by beinvisible** [comment]
Please read carefully, I am talking specifically about period = '**week**'.

btw. I have a pull request coming.

**Comment by larcorba** [comment]
Gotcha, thanks! The initial script wasn't written by us, so after reading the script a bit further I understand what you're coming from. I was not aware there was an option included for a 'week' period. Would have to test this again someday to reproduce the issue with that feature.

I'll check out your pull request once it's ready, thanks for contributing :)

**Workaround: For now the script is usable by keeping the _daily_ partition for history and _monthly_ for trends.**

**Comment by beinvisible** [comment]
while we're at it: the weekly partition names are a bit odd as well. 'Day' and 'month' follow ISO8601, but 'week' doesnt. But changing this would break existing installations using weekly partitioning. I have included this as separate commit.

**Comment by larcorba** [comment]
Looks like you've made some nice improvements to the script. Thanks! If could create a seperate pull request for the ISO standard change and the fixing of the bug I can review and merge them seperatly. 

This way I can keep the versioning clean for people still wanting to use weekly, but not wanting to use the new ISO formated weekly partitioning.

**Comment by beinvisible** [comment]
that was my original plan actually, but github adds all commits from the same branch to the PR... :-P

**Comment by beinvisible** [comment]
Ok, I removed the last commit from the main branch. Will do a separate PR for the timestamp.

**Comment by larcorba** [comment]
Thanks for splitting them! I will test the script in some environments a.s.a.p. and merge your changes if everything works well.

**Comment by larcorba** [comment]
Hi Beinvisible,

I checked the functionality over the past week and this weekend I did some final code checks. Last but not least, this monday morning I was able to verify if the delete partition part was still working.

Your code looks great. Creation of partitions, check if present and deletion are all working (Daily, weekly and monthly). Thank you for your contribution, the Zabbix community will be very pleased!

The last part I would like to pickup is writing a blog post on blog.zabbix.com or editing my current one to also include weekly partitioning. But I don't know if I'll have time in the near future, so that will have to wait a bit. For now I added a small part in the Readme.

Once again thank you for the contribution and have a great week.

Regards,

Nathan
